lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -488,19 +498,90 @@ getLegacyTransaction tid = do
     let wits = mempty
     pure (Tx tid inps outs, wits)
 
+
+putWitness :: TxWitness -> Put
+putWitness (TxWitness bytes) = putByteString bytes
+
 putSignedTx :: [(TxIn, Coin)] -> [TxOut] -> [TxWitness] -> Put
 putSignedTx inputs outputs witnesses = do
-    putTx inputs outputs
+    putIO inputs outputs
     unless (length inputs == length witnesses) $
         fail "number of witnesses must equal number of inputs"
     mapM_ putWitness witnesses
+
+withBytes :: forall a e. (ByteString -> a) -> ExceptT e PutM () -> ExceptT e PutM a
+withBytes act putData = act <$> mapExceptT runExceptTPut putData
+
+withHash :: ExceptT e PutM () -> ExceptT e PutM ByteString
+withHash = withBytes blake2b256
+
+runExceptTPut :: PutM (Either e ()) -> PutM (Either e ByteString)
+runExceptTPut x = do
+    x' <- x
+    case x' of
+        Left e -> return $ Left e
+        Right () -> do
+            let bs = BL.toStrict $ runPut (fmap (const ()) x)
+            return $ Right bs
+
+runEPut :: PutM (Either e ()) -> Either e ByteString
+runEPut x =
```

Comment:
I'm not quite sure what I'm doing with `runEPut` and `runExceptTPut`. It's very ugly. I'm sure there's a better way.